ZividOpenCV/CaptureUndistortRGB/CaptureUndistortRGB.cpp

```diff
@@ -0,0 +1,129 @@
+
+        std::cout << "Capturing a frame" << std::endl;
+        const auto frame = camera.capture();
+
+        const auto intrinsics = camera.intrinsics();
+
+        std::cout << "Setting up visualization" << std::endl;
+        Zivid::CloudVisualizer vis;
+        zivid.setDefaultComputeDevice(vis.computeDevice());
+
+        std::cout << "Displaying the point cloud" << std::endl;
+        vis.showMaximized();
+        vis.show(frame);
+        vis.resetToFit();
+
+        std::cout << "Running the visualizer. Blocking until the window closes" << std::endl;
+        vis.run();
+
+        std::cout << "Converting ZDF point cloud to OpenCV format" << std::endl;
+
+        const auto pointCloud = frame.getPointCloud();
+        cv::Mat rgb((int)pointCloud.height(), (int)pointCloud.width(), CV_8UC3, cv::Scalar(0, 0, 0));
+        pointCloudToRGB(pointCloud, rgb);
+
+        std::cout << "Undistorting the RGB image" << std::endl;
+
+        cv::Size size = rgb.size();
+
+        cv::Mat rgbUndistorted;
+        cv::Mat rgbUndistortedFull;
+
+        cv::Mat distortionCoefficients(cv::Size(1, 5), CV_64FC1, cv::Scalar(0));
+        cv::Mat cameraMatrix(cv::Size(3, 3), CV_64FC1, cv::Scalar(0));
+        cv::Mat optimalCameraMatrix;
+
+        getCameraIntrinsics(camera, distortionCoefficients, cameraMatrix);
+
+        optimalCameraMatrix = cv::getOptimalNewCameraMatrix(cameraMatrix, distortionCoefficients, size, 1, size);
+
+        cv::undistort(rgb, rgbUndistorted, cameraMatrix, distortionCoefficients);
+        cv::undistort(rgb, rgbUndistortedFull, cameraMatrix, distortionCoefficients, optimalCameraMatrix);
+
+        std::cout << "Displaying and saving the RGB image" << std::endl;
+
+        displayRGB(rgb, "Distorted RGB image");
+        cv::imwrite("Distorted RGB image.jpg", rgb);
+
+        std::cout << "Displaying and saving the undistorted RGB image" << std::endl;
+
+        displayRGB(rgbUndistorted, "Undistorted RGB image");
+        cv::imwrite("Undistorted RGB image.jpg", rgbUndistorted);
+
+        std::cout << "Displaying and saving the Undistorted RGB image - full" << std::endl;
+
+        displayRGB(rgbUndistortedFull, "Undistorted RGB image - full");
+        cv::imwrite("Undistorted RGB image - full.jpg", rgbUndistorted);
+    }
+    catch(const std::exception &e)
+    {
+        std::cerr << "Error: " << Zivid::toString(e) << std::endl;
+        return EXIT_FAILURE;
+    }
+}
+
+void pointCloudToRGB(const Zivid::PointCloud &pointCloud, cv::Mat &rgb)
+{
+    for(int i = 0; i < pointCloud.height(); i++)
+    {
+        for(int j = 0; j < pointCloud.width(); j++)
+        {
+            cv::Vec3b &color = rgb.at<cv::Vec3b>(i, j);
+            color[0] = pointCloud(i, j).blue();
+            color[1] = pointCloud(i, j).green();
+            color[2] = pointCloud(i, j).red();
+        }
+    }
+}
+
+void getCameraIntrinsics(const Zivid::Camera &camera, cv::Mat &distortionCoefficients, cv::Mat &cameraMatrix)
```

Comment:
We should avoid out parameters here too, and initialize the variables inside the function and then return them.

One alternative is to return both values using a struct or tuple, like the [C++ Core Guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f21-to-return-multiple-out-values-prefer-returning-a-struct-or-tuple) suggest when having to  return several values.

An alternative would be to split this in two functions, one for creating `distortionCoefficients` and one for creating `cameraMatrix`, then you only need one return value for each function.

In either case, the resulting variables in `main()` should be made `const`.

You can also consider passing `camera.intrinsics()` or `camera.intrinsics().distortion()` into the function(s) instead of passing `camera`. I think this will make the code a bit cleaner since you don't have to keep calling all those functions.

Reply:
Done with a tuple and passing camera.intrinsics(). This is fun.

Reply:
Nice!